### PR TITLE
feat: reset history in useHistoryTravel

### LINKED ISSUE #517: useHistoryTravel is missing reset method
At the moment, there is no reset method to reset history.

### PR DESCRIPTION
**Pull request by jinixx**
Method to reset history to initial value or with a new initial value.

fixes #517

### CODE REVIEW

**Comment by brickspert** [comment]
You should provide an example for reset.

**Comment by jinixx** [comment]
> You should provide an example for reset.

Example and demo added.

File: packages/hooks/src/useHistoryTravel/index.ts [hunk 1]
```diff
@@ -37,6 +37,24 @@ export default function useHistoryTravel<T>(initialValue?: T) {
...
+
+      setHistory({
+        present: _initial,
+        future: [],
+        past: []
+      });
+    },
+    [history, setHistory]
+  );
+
```
Comment (brickspert) [inline packages/hooks/src/useHistoryTravel/index.ts]:
```suggestion

  const initialValueRef = useRef(initialValue);

  const reset = useCallback(
    (newInitialValue?: T) => {
      setHistory({
        present: newInitialValue === undefined ? initialValueRef.current : newInitialValue,
        future: [],
        past: []
      });
    },
    [setHistory]
  );
```

File: packages/hooks/src/useHistoryTravel/demo/demo1.tsx [hunk 2]
```diff
@@ -56,6 +62,10 @@ export default () => {
           {' '}
           Redo{' '}
         </button>
+        <button type="button" disabled={!backLength && !forwardLength} onClick={onReset}>
```
Comment (brickspert) [inline packages/hooks/src/useHistoryTravel/demo/demo1.tsx:65]: 应该和前面的按钮有个边距

**Review by brickspert — CHANGES_REQUESTED** [review]

File: packages/hooks/src/useHistoryTravel/index.ts (hunk 1 above)
Comment (jinixx) [reply packages/hooks/src/useHistoryTravel/index.ts]: without this, how can I implement a similar feature if the game board layout has changed?

```javascript
const initialBoardData = useMemo(() => getBoardLayout(config), [config]);

const {
    value,
    setValue,
    backLength,
    forwardLength,
    back,
    forward,
    go,
    reset
} = useHistoryTravel({
    ...initialStateForTimeTravel,
    boardData: initialBoardData
});
```

File: packages/hooks/src/useHistoryTravel/demo/demo1.tsx (hunk 2 above)
Comment (jinixx) [reply packages/hooks/src/useHistoryTravel/demo/demo1.tsx:65]: Sorry, I don't understand this.

File: packages/hooks/src/useHistoryTravel/index.ts (hunk 1 above)
Comment (brickspert) [reply packages/hooks/src/useHistoryTravel/index.ts]: I have modified the suggested code~~

File: packages/hooks/src/useHistoryTravel/demo/demo1.tsx (hunk 2 above)
Comment (brickspert) [reply packages/hooks/src/useHistoryTravel/demo/demo1.tsx:65]: Sorry，I mean we need to add a margin between the buttons

File: packages/hooks/src/useHistoryTravel/index.ts (hunk 1 above)
Comment (jinixx) [reply packages/hooks/src/useHistoryTravel/index.ts]: @brickspert your suggestion is similar to my first attempt. But, I amended it after I realised from the tests that undefined is a valid initialValue. Reset should allow setting initialValue to undefined as well.

Comment (brickspert) [reply packages/hooks/src/useHistoryTravel/index.ts]: reset is a bit troublesome to set two parameters  for one thing.

 or you can use like this

```js
const reset = (...params)=>{
   if(params.length > 0){
      ...
   }else{
      ...
   }
}
```

in this code , `reset()` is not equal to `reset(undefined)`.

**Review by brickspert — APPROVED** [review]
LGTM

**Review by ttys026 — APPROVED** [review]
LGTM

**Pull request merged**